librz/bin/p/bin_elf.inc

```diff
@@ -808,302 +622,578 @@ static RzBinReloc *reloc_convert(struct Elf_(rz_bin_elf_obj_t) * bin, RzBinElfRe
+	rz_return_val_if_fail(bf && bf->o && bf->o->bin_obj, NULL);
+
+	RzBinAddr *ptr = NULL;
+	struct rz_bin_elf_symbol_t *symbol;
+	int i;
+
+	struct Elf_(rz_bin_elf_obj_t) *obj = bf->o->bin_obj;
+	RzList *ret = rz_list_newf((RzListFree)free);
+	if (!ret) {
+		return NULL;
+	}
+	ut64 pa = Elf_(rz_bin_elf_get_entry_offset)(obj);
+	if (pa != UT64_MAX) {
+		if (!(ptr = RZ_NEW0(RzBinAddr))) {
+			return ret;
+		}
+		ptr->paddr = pa;
+		ptr->vaddr = Elf_(rz_bin_elf_p2v)(obj, ptr->paddr);
+		ptr->hpaddr = 0x18; // e_entry offset in ELF header
+		ptr->hvaddr = UT64_MAX;
+
+		if (ptr->vaddr != (ut64)obj->ehdr.e_entry && Elf_(rz_bin_elf_is_executable)(obj)) {
+			eprintf("Cannot determine entrypoint, using 0x%08" PFMT64x ".\n", ptr->vaddr);
+		}
+
+		if (bf->o->sections) {
+			RzListIter *iter;
+			RzBinSection *section;
+			rz_list_foreach_prev(bf->o->sections, iter, section) {
+				if (!strcmp(section->name, "ehdr")) {
+					ptr->hvaddr = section->vaddr + ptr->hpaddr;
+					break;
+				}
 			}
 		}
-		break;
-	}
-	case EM_X86_64: {
-		int word = 0;
-		switch (rel->type) {
-		case RZ_X86_64_8:
-			word = 1;
-			val = S + A;
-			break;
-		case RZ_X86_64_16:
-			word = 2;
-			val = S + A;
-			break;
-		case RZ_X86_64_32:
-		case RZ_X86_64_32S:
-			word = 4;
-			val = S + A;
-			break;
-		case RZ_X86_64_64:
-			word = 8;
-			val = S + A;
-			break;
-		case RZ_X86_64_GLOB_DAT:
-		case RZ_X86_64_JUMP_SLOT:
-			word = 4;
-			val = S;
-			break;
-		case RZ_X86_64_PC8:
-			word = 1;
-			val = S + A - P;
-			break;
-		case RZ_X86_64_PC16:
-			word = 2;
-			val = S + A - P;
-			break;
-		case RZ_X86_64_PC32:
-			word = 4;
-			val = S + A - P;
-			break;
-		case RZ_X86_64_PC64:
-			word = 8;
-			val = S + A - P;
-			break;
-		case RZ_X86_64_PLT32:
-			word = 4;
-			val = L + A - P;
-			break;
-		case RZ_X86_64_RELATIVE:
-			word = 8;
-			val = B + A;
-			break;
-		default:
-			//eprintf ("relocation %d not handle at this time\n", rel->type);
-			break;
+		if (ptr->hvaddr == UT64_MAX) {
+			Elf_(rz_bin_elf_p2v_new)(obj, ptr->hpaddr);
 		}
-		switch (word) {
-		case 0:
-			break;
-		case 1:
-			buf[0] = val;
-			iob->write_at(iob->io, rel->rva, buf, 1);
-			break;
-		case 2:
-			rz_write_le16(buf, val);
-			iob->write_at(iob->io, rel->rva, buf, 2);
-			break;
-		case 4:
-			rz_write_le32(buf, val);
-			iob->write_at(iob->io, rel->rva, buf, 4);
-			break;
-		case 8:
-			rz_write_le64(buf, val);
-			iob->write_at(iob->io, rel->rva, buf, 8);
-			break;
+
+		if (obj->ehdr.e_machine == EM_ARM) {
+			int bin_bits = Elf_(rz_bin_elf_get_bits)(obj);
+			if (bin_bits != 64) {
+				ptr->bits = 32;
+				if (ptr->vaddr & 1) {
+					ptr->vaddr--;
+					ptr->bits = 16;
+				}
+				if (ptr->paddr & 1) {
+					ptr->paddr--;
+					ptr->bits = 16;
+				}
+			}
 		}
-		break;
+		rz_list_append(ret, ptr);
+	}
+
+	// add entrypoint for jni libraries
+	// NOTE: this is slow, we shouldnt find for java constructors here
+	if (!(symbol = Elf_(rz_bin_elf_get_symbols)(obj))) {
```

Comment:
Since you are the Java expert, @wargio please take a look at this particular piece of code.

Reply:
LGTM